Add validation for VK_KHR_shader_float_controls

AttilioProvenzano-ARM opened the pull request:
This adds core validation for VK_KHR_shader_float_controls.
It depends on validator support in SPIRV-Tools: https://github.com/KhronosGroup/SPIRV-Tools/pull/2177

It checks that the shader submitted to vkCreateShaderModule respects
the enabled features when it sets the execution mode for floats.
It also adds relevant entries to the automated capability checks.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/Vulkan-ValidationLayers?pullRequest=513) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/Vulkan-ValidationLayers?pullRequest=513) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/Vulkan-ValidationLayers?pullRequest=513) it.</sub>

tobine commented:
@AttilioProvenzano-ARM Can you please rebase off of latest master and then verify that CI tests run cleanly?

AttilioProvenzano-ARM commented:
> @AttilioProvenzano-ARM Can you please rebase off of latest master and then verify that CI tests run cleanly?

@tobine I rebased off of latest master, This PR depends on https://github.com/KhronosGroup/SPIRV-Tools/pull/2177 though, so I'm not expecting the CI to succeed until that is merged.

AttilioProvenzano-ARM commented:
Hi @jzulauf-lunarg, this is still active, it was blocked on a discussion regarding the split of responsibilities between this and SPIRV-Tools.
A decision was made yesterday (https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/674), so I can keep working on this.

AttilioProvenzano-ARM commented:
Hi, just a quick update. We've been really busy with GDC preparations and I couldn't look into it, sorry about that!

I'll aim at completing this right after GDC, end of March/beginning of April. :)

AttilioProvenzano-ARM commented:
I just updated this PR, the checks are now moved from SpirvTools to the layers (see https://github.com/KhronosGroup/SPIRV-Tools/pull/2177 for the previous SpirvTools PR).

 @tobine @chrisforbes could you please review it?

jzulauf-lunarg commented:
@greg-lunarg -- can you give the shader validation section a review?

jzulauf-lunarg commented:
@chrisforbes  -- is it possible for you to give this a look?

mikes-lunarg commented on layers/shader_validation.cpp:
```diff
@@ -1827,6 +1845,137 @@ bool CoreChecks::ValidateShaderStageInputOutputLimits(shader_module const *src,
+                    if (found == first_denorm_execution_mode.end()) {
+                        first_denorm_execution_mode[bit_width] = static_cast<spv::ExecutionMode>(mode);
+                    } else if (found->second != mode && !GetEnabledFeatures()->float_controls.separateDenormSettings) {
+                        skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0,
```
I'm not sure this is quite right.  By using `bit_width` as the map key you will not detect separate denorm settings for different widths as the error message states.  Same for rounding settings below.

jzulauf-lunarg replied:
Not sure you need a map at all if you store {width, mode} of the first, and just compare all subsequent against it.

AttilioProvenzano-ARM replied:
Agreed with both, the checks were wrong. I fixed them in b02dc30e81d3afca747550c58467f75bbcc46c78, thanks for your feedback! 👍

AttilioProvenzano-ARM commented:
@mikes-lunarg Thanks for reviewing, I fixed the error with `separate*` validation.

Regarding the test cases, I considered porting the ones I had made for SPIRV-Tools before moving the checks to the layers, but as you said the extension is not widely available yet, so I wouldn't be able to run them.
